### PR TITLE
Fix Dockerfile

### LINKED ISSUE #31: Errors when violation of no-invalid-template-strings rule
The `codeclimate-tslint` engine analysis errors when:

* It is configured to check for violations of the `no-invalid-template-strings` rule
* The code has a violation of the check

The engine actually outputs the error in JSON format, in the format of a static analysis issue, which itself is probably not right.

The salient error it reports is "Error: rule named no-invalid-template-strings is not found."

I've created a repository which replicates this bug here: https://github.com/codeclimate-testing/replicate-tslint-bug

### PR DESCRIPTION
**Pull request by pointlessone**
There was a possibility to improperly build an image that would contain docs from mismatched TSLint version. This could result in error caused by the engine being unable to find issue descriptions. This appears to be the reason current beta channel of this engine is broken — it contains docs for TSLint 5.0 while 5.1 is used in code.

This Dockerfile eliminates the issue and adds a few clean up steps to bring the image size a little down.

Fixes #31

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,32 +1,35 @@
...
 COPY ./bin/ ./bin/
 
-RUN npm install --global yarn && \
+RUN \
+  : Install needed tools && \
+  npm install --global yarn && \
   apk --update add git jq && \
-  bin/get-tslint-rules "$VER_TSLINT" && \
-  cat engine.json | jq ".version = \"$VER_TSLINT\"" > /tmp/engine.json && \
+  : Install NPM packages && \
```
Comment (dblandin) [inline Dockerfile]: Do we need these comments? The commands seem fairly revealing on their own.

File: engine.json [hunk 2]
```diff
@@ -7,5 +7,5 @@
   },
   "languages" : ["TypeScript"],
   "version": "51e260eb5c8842431dad261fa70a700382cf9e60",
-  "spec_version": "0.2.3"
+  "spec_version": "0.3.1"
```
Comment (dblandin) [inline engine.json]: Did anything else in this PR change the spec compatibility? If not, this bump could probably be its own commit.

File: Dockerfile (hunk 1 above)
Comment (pointlessone) [reply Dockerfile]: No, not really. I added them to the original chain to annotate it. It was a bit more cryptic.

File: engine.json (hunk 2 above)
Comment (pointlessone) [reply engine.json]: `engine.json` has been moved from `/usr/src/app` to `/`.

Comment (dblandin) [reply engine.json]: Ahh, awesome 👍. Still think that might be nice to specify in a separate commit.

File: Dockerfile [hunk 3]
```diff
@@ -1,32 +1,35 @@
...
+   yarn install && \
+  : Update engine.json &&\
+  jq <engine.json ".version = \"$(bin/version tslint)\"" > /engine.json && \
```
Comment (dblandin) [inline Dockerfile]: This is really cool! Wasn't aware of this invocation style 👌

**Comment by tkqubo** [comment]
Sorry for my merge being so terribly late!  But this PR fails in my environment, with the console saying that the user has no permission to perform `rm -rfv dist`.

screen shot 0029-10-02 at 20 01 28

After moving `USER app` directive *after* `RUN npm run build`, docker build worked successfully.

## BEFORE

```
USER app

COPY . ./
RUN npm run build

VOLUME /code
WORKDIR /code
```

## AFTER

```
COPY . ./
RUN npm run build

USER app

VOLUME /code
WORKDIR /code
```

Is that OK to move `USER` directive? @pointlessone

**Comment by tkqubo** [comment]
Thanks, I followed you advice, but using `chown` instead 😄
